processing/src/main/java/org/apache/druid/query/UnionDataSource.java

```diff
@@ -52,6 +57,51 @@ public UnionDataSource(@JsonProperty("dataSources") List<TableDataSource> dataSo
     return dataSources;
   }
 
+  @Override
+  public List<DataSource> getChildren()
+  {
+    return ImmutableList.copyOf(dataSources);
+  }
+
+  @Override
+  public DataSource withChildren(List<DataSource> children)
+  {
+    if (children.size() != dataSources.size()) {
```

Comment:
why must these sizes match?

Reply:
It's a precondition for the method, since its intent is for replacing inputs with inlined/etc versions, so the number of children shouldn't change. I'll add a comment in the interface.